src/matrix_cookbook/mat_eq27_lib.lean

```diff
@@ -0,0 +1,206 @@
+  A 0 3*A 2 0*A 3 2 - A 0 3*A 2 2*A 3 0 + A 1 1*A 2 2*A 3 3 - A 1 1*A 2 3*A 3 2 - A 1 2*A 2 1*A 3 3 + A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1
+:= 
+begin
+  rw det4_one_add_a, rw det4, ring,
+end
+
+lemma trace_a_all_squared {A : matrix (fin 4) (fin 4) R}:
+(trace A)^2 = 
+A 0 0^2 + A 1 1^2 + A 2 2^2 + A 3 3^2 + 
+2*A 0 0*A 1 1 + 2*A 0 0*A 2 2 + 2*A 0 0*A 3 3 + 
+2*A 1 1*A 2 2 + 2*A 1 1*A 3 3 + 2*A 2 2*A 3 3 
+:= begin
+  rw trace_a_fin4, rw pow_two, ring,
+end
+
+lemma eq_27_rhs_part1{A : matrix (fin 4) (fin 4) R}:
+(trace A)^3 - 3*trace A * trace (A^2) + 2 * trace (A^3) = 6*(A 0 0*A 1 1*A 2 2 - A 0 0*A 1 2*A 2 1 - A 0 1*A 1 0*A 2 2 + A 0 1*A 1 2*A 2 0 + A 0 2*A 1 0*A 2 1 - A 0 2*A 1 1*A 2 0 + A 0 0*A 1 1*A 3 3 - 
+A 0 0*A 1 3*A 3 1 - A 0 1*A 1 0*A 3 3 + A 0 1*A 1 3*A 3 0 + A 0 3*A 1 0*A 3 1 - A 0 3*A 1 1*A 3 0 + A 0 0*A 2 2*A 3 3 - A 0 0*A 2 3*A 3 2 - 
+A 0 2*A 2 0*A 3 3 + A 0 2*A 2 3*A 3 0 + A 0 3*A 2 0*A 3 2 - A 0 3*A 2 2*A 3 0 + A 1 1*A 2 2*A 3 3 - A 1 1*A 2 3*A 3 2 - A 1 2*A 2 1*A 3 3 + 
+A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1) := 
+begin
+  rw trace_a_fin4, rw trace_a_squared, rw trace_a_cubed, ring,
+end
+
+lemma eq_27_rhs_part2{A : matrix (fin 4) (fin 4) R}:
+(trace A)^2 - trace (A^2) = 
+  2*(A 0 0*A 1 1 - A 0 1*A 1 0 + A 0 0*A 2 2 - A 0 2*A 2 0 + A 0 0*A 3 3 - A 0 3*A 3 0 + 
+A 1 1*A 2 2 - A 1 2*A 2 1 + A 1 1*A 3 3 - A 1 3*A 3 1 + A 2 2*A 3 3 - A 2 3*A 3 2) := 
+begin 
+  rw trace_a_all_squared, rw trace_a_squared, ring,
+end
+
+
+lemma eq_27_rhs {A : matrix (fin 4) (fin 4) R}:
+trace A + (1/2)*( (trace A)^2 - trace (A^2)) + 
+              (1/6)*( (trace A)^3 - 3*trace A * trace (A^2) + 2 * trace (A^3) ) = 
+ A 0 0 + A 1 1 + A 2 2 + A 3 3 + 
+  A 0 0*A 1 1 - A 0 1*A 1 0 + A 0 0*A 2 2 - A 0 2*A 2 0 + A 0 0*A 3 3 - A 0 3*A 3 0 + A 1 1*A 2 2 - A 1 2*A 2 1 + 
+  A 1 1*A 3 3 - A 1 3*A 3 1 + A 2 2*A 3 3 - A 2 3*A 3 2 + 
+  A 0 0*A 1 1*A 2 2 - A 0 0*A 1 2*A 2 1 - A 0 1*A 1 0*A 2 2 + A 0 1*A 1 2*A 2 0 + A 0 2*A 1 0*A 2 1 - A 0 2*A 1 1*A 2 0 + A 0 0*A 1 1*A 3 3 - A 0 0*A 1 3*A 3 1 - 
+  A 0 1*A 1 0*A 3 3 + A 0 1*A 1 3*A 3 0 + A 0 3*A 1 0*A 3 1 - A 0 3*A 1 1*A 3 0 + A 0 0*A 2 2*A 3 3 - A 0 0*A 2 3*A 3 2 - A 0 2*A 2 0*A 3 3 + A 0 2*A 2 3*A 3 0 + 
+  A 0 3*A 2 0*A 3 2 - A 0 3*A 2 2*A 3 0 + A 1 1*A 2 2*A 3 3 - A 1 1*A 2 3*A 3 2 - A 1 2*A 2 1*A 3 3 + A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1
+:= 
+begin
+  have h2: (2:R) ≠ 0, {norm_num,},
+  have h6: (6:R) ≠ 0, {norm_num,},
+  rw eq_27_rhs_part1, rw ← mul_assoc, rw one_div_mul_cancel h6, 
+  rw one_mul,
+  rw eq_27_rhs_part2, rw ← mul_assoc, rw one_div_mul_cancel h2, 
+  rw one_mul,
+  rw trace_a_fin4, ring,
+end
+
+
+
+lemma eq_27_before_last {A : matrix (fin 4) (fin 4) R} :
+  det (1 + A) - det A - 1 = trace A + 
+    (1/2)*( (trace A)^2 - trace (A^2)) + 
+    (1/6)*( (trace A)^3 - 3*trace A * trace (A^2) + 2 * trace (A^3) ) := 
+begin
+  rw eq_27_lhs, rw eq_27_rhs,
+end
```

Comment:
You could combine these two into a single proof as something like
```suggestion
begin
  transitivity  A 0 0 + A 1 1 + A 2 2 + A 3 3 + 
  A 0 0*A 1 1 - A 0 1*A 1 0 + A 0 0*A 2 2 - A 0 2*A 2 0 + A 0 0*A 3 3 - A 0 3*A 3 0 + A 1 1*A 2 2 - A 1 2*A 2 1 + 
  A 1 1*A 3 3 - A 1 3*A 3 1 + A 2 2*A 3 3 - A 2 3*A 3 2 + 
  A 0 0*A 1 1*A 2 2 - A 0 0*A 1 2*A 2 1 - A 0 1*A 1 0*A 2 2 + A 0 1*A 1 2*A 2 0 + A 0 2*A 1 0*A 2 1 - A 0 2*A 1 1*A 2 0 + A 0 0*A 1 1*A 3 3 - A 0 0*A 1 3*A 3 1 - 
  A 0 1*A 1 0*A 3 3 + A 0 1*A 1 3*A 3 0 + A 0 3*A 1 0*A 3 1 - A 0 3*A 1 1*A 3 0 + A 0 0*A 2 2*A 3 3 - A 0 0*A 2 3*A 3 2 - A 0 2*A 2 0*A 3 3 + A 0 2*A 2 3*A 3 0 + 
  A 0 3*A 2 0*A 3 2 - A 0 3*A 2 2*A 3 0 + A 1 1*A 2 2*A 3 3 - A 1 1*A 2 3*A 3 2 - A 1 2*A 2 1*A 3 3 + A 1 2*A 2 3*A 3 1 + A 1 3*A 2 1*A 3 2 - A 1 3*A 2 2*A 3 1,
  { -- put your lhs proof here
    sorry },
  { -- put your rhs proof here,
    sorry },
```

Reply:
You mean get rid of the lemmas eq_27_lhs and eq_27_rhs and put them all in the before last lemma?

Reply:
Yes, exactly!